examples/component.rs

```diff
@@ -0,0 +1,166 @@
+extern crate env_logger;
+extern crate log;
+extern crate z3;
+
+use z3::{
+    ast::{Ast, Bool, Datatype, Int, BV},
+    Config, Context, DatatypeBuilder, DatatypeSort, SatResult, Solver,
+};
+
+fn main() {
+    let _ = env_logger::try_init();
+    let cfg = Config::new();
+    let ctx = Context::new(&cfg);
+
+    // Define a new datatype named Instruction which has 3 variants.
+    // 1. LocalGet 0
+    // 2. I32Add
+    // 3. Nop
+    let components = DatatypeBuilder::new(&ctx)
+        .variant("LocalGet0", &[])
+        .variant("I32Add", &[])
+        .variant("Nop", &[])
+        .finish("Instruction");
+
+    // Function that converts an Instruction datatype into z3::ast::Int which represents the number
+    // of values the instruction pops off from the stack.
+    fn pop_cnts<'a>(
+        ctx: &'a Context,
+        components: &DatatypeSort<'a>,
+        instr: &Datatype<'a>,
+    ) -> Int<'a> {
+        components.variants[0]
+            .tester
+            .apply(&[&instr.clone().into()])
+            .as_bool()
+            .unwrap()
+            .ite(
+                &(Int::from_u64(&ctx, 0)),
+                &components.variants[1]
+                    .tester
+                    .apply(&[&instr.clone().into()])
+                    .as_bool()
+                    .unwrap()
+                    .ite(&(Int::from_u64(&ctx, 2)), &(Int::from_u64(&ctx, 0))),
+            )
+    }
+
+    // Function that converts an Instruction datatype into z3::ast::Int which represents the number
+    // of values the instruction pushes to the stack.
+    fn push_cnts<'a>(
+        ctx: &'a Context,
+        components: &DatatypeSort<'a>,
+        instr: &Datatype<'a>,
+    ) -> Int<'a> {
+        components.variants[0]
+            .tester
+            .apply(&[&instr.clone().into()])
+            .as_bool()
+            .unwrap()
+            .ite(
+                &(Int::from_u64(&ctx, 1)),
+                &components.variants[1]
+                    .tester
+                    .apply(&[&instr.clone().into()])
+                    .as_bool()
+                    .unwrap()
+                    .ite(&(Int::from_u64(&ctx, 1)), &(Int::from_u64(&ctx, 0))),
+            )
+    }
+
+    // Well formed program constraint, given a slice of instructions.
+    fn is_valid<'a>(
+        ctx: &'a Context,
+        components: &DatatypeSort<'a>,
+        instrs: &[Datatype<'a>],
+    ) -> Bool<'a> {
+        let cnt = Int::from_u64(&ctx, 0);
+        let zero = Int::from_u64(&ctx, 0);
+
+        fn is_valid_helper<'a>(
```

Comment:
Can we un-nest this function so that it resides outside is_valid? It's a bit confusing to me to see a function defined inside another one. Also, a comment on top of is_valid_helper explaining what it does would be helpful.

Reply:
All functions that were nested in main() have been moved outside of the main. The reason I nested `is_valid_helper` within `is_valid` is to provide cleaner API and hide `is_valid_helper`. 

Hope my comments are useful. I just added line by line comments for your understanding for `is_valid_helper`